File: native/android-app/src/main/java/com/fire/app/session/FireCookieSelfHealingCoordinator.kt
package com.fire.app.session

import android.os.Looper
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeoutOrNull
import uniffi.fire_uniffi_session.CookieSelfHealingHandler
import uniffi.fire_uniffi_session.CookieSelfHealingPhaseState
import uniffi.fire_uniffi_session.CookieSelfHealingRequestState
import uniffi.fire_uniffi_session.CookieSelfHealingResultState

class FireCookieSelfHealingRuntimeHandler(
    private val sessionStore: FireSessionStore,
) : CookieSelfHealingHandler {
    override fun healCookies(
        request: CookieSelfHealingRequestState,
    ): CookieSelfHealingResultState {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            return CookieSelfHealingResultState(
                completed = false,
                sessionEpoch = request.sessionEpoch,
            )
        }

        return runBlocking {
            val coordinator = FireWebViewLoginCoordinator(sessionStore)
            val completed = withTimeoutOrNull(15_000L) {
                runCatching {
                    when (request.phase) {
                        CookieSelfHealingPhaseState.SWEEP -> {
                            coordinator.sweepCookies(
                                names = request.cookieNames,
                                targetUrl = request.targetUrl,
                            )
                        }
                        CookieSelfHealingPhaseState.NUCLEAR_RESET -> {
                            coordinator.nuclearResetCookies(
                                targetUrl = request.targetUrl,
                            )
                        }
                    }
                }.isSuccess
            } ?: false
            CookieSelfHealingResultState(
                completed = completed,
                sessionEpoch = request.sessionEpoch,
            )
        }
    }
}

File: native/ios-app/Sources/FireAppSession/FireCookieSelfHealingCoordinator.swift
import Foundation

final class FireCookieSelfHealingRuntimeHandler: CookieSelfHealingHandler, @unchecked Sendable {
    private let loginCoordinator: FireWebViewLoginCoordinator

    init(loginCoordinator: FireWebViewLoginCoordinator) {
        self.loginCoordinator = loginCoordinator
    }

    nonisolated func healCookies(
        request: CookieSelfHealingRequestState
    ) -> CookieSelfHealingResultState {
        if Thread.isMainThread {
            return CookieSelfHealingResultState(
                completed: false,
                sessionEpoch: request.sessionEpoch
            )
        }

        let semaphore = DispatchSemaphore(value: 0)
        let result = LockedCookieSelfHealingResultState(
            CookieSelfHealingResultState(
                completed: false,
                sessionEpoch: request.sessionEpoch
            )
        )
        DispatchQueue.main.async { [loginCoordinator] in
            Task { @MainActor in
                do {
                    let targetURL = URL(string: request.targetUrl)
                    switch request.phase {
                    case .sweep:
                        _ = try await loginCoordinator.sweepCookies(
                            names: request.cookieNames,
                            targetURL: targetURL
                        )
                    case .nuclearReset:
                        try await loginCoordinator.nuclearResetCookies(
                            targetURL: targetURL
                        )
                    }
                    result.set(
                        CookieSelfHealingResultState(
                            completed: true,
                            sessionEpoch: request.sessionEpoch
                        )
                    )
                } catch {
                    result.set(
                        CookieSelfHealingResultState(
                            completed: false,
                            sessionEpoch: request.sessionEpoch
                        )
                    )
                }
                semaphore.signal()
            }
        }
        _ = semaphore.wait(timeout: .now() + 15)
        return result.get()
    }
}

private final class LockedCookieSelfHealingResultState: @unchecked Sendable {
    private let lock = NSLock()
    private var value: CookieSelfHealingResultState

    init(_ value: CookieSelfHealingResultState) {
        self.value = value
    }

    func set(_ value: CookieSelfHealingResultState) {
        lock.lock()
        self.value = value
        lock.unlock()
    }

    func get() -> CookieSelfHealingResultState {
        lock.lock()
        defer { lock.unlock() }
        return value
    }
}
